### PR TITLE
Another "big" pull request which fixes tons of bugs

### PR DESCRIPTION
**Pull request by 1951FDG**
Hi, I wasn't expecting to contribute this soon again, :)

The commits should be self explanatory, but if you have any questions let me know. Most of the bugs are pretty though to replicate, hopefully the commit messages contain enough info for if you desire to test before and after, sometimes you may need to disable the Handler run in CrashlyticApplication...

I've had to set Kotlin `jvmTarget = '1.8'`  for cc8da7f, since `androidx.core:core-ktx` depends on it. For some reason my Android Studio was using `1.6` as `jvmTarget` for `Kotlin` per default, so I've updated all Gradle files to explicitly set the  `jvmTarget` for `Kotlin`.

Recently I was testing states in my own application, as in recreating activities, forcing activity recreation with dev option, `Don't keep activities`, which destroys activities as the user leaves it, to simulate when the system decides to destroy an Activity for whatever reason, which usually happens when the user leaves the activity for a while...

Recommendations for future improvements, implement saving of log filter state, currently, log filter state is not saved and not restored if Activity is recreated by the system.

References used, just in case, so you don't need Google...

[Fragment Lifecycles in the Age of Jetpack](https://zsmb.co/fragment-lifecycles-in-the-age-of-jetpack/) - See section titled, `LifecycleOwners and observing LiveData`

[LiveData with SnackBar, Navigation and other events (the SingleLiveEvent case)](https://medium.com/androiddevelopers/livedata-with-snackbar-navigation-and-other-events-the-singleliveevent-case-ac2622673150)

[Restore RecyclerView scroll position](https://medium.com/androiddevelopers/restore-recyclerview-scroll-position-a8fbdc9a9334)

### CODE REVIEW

**Comment by hannesa2** [comment]
Thank you !

I know this wish (not a request !) is maybe too much, but it would be super nice, to have a test in advance which fails before the fix and the fix solves it.  
eg. the duplicate last line I never observed

File: LogcatCoreLib/src/main/java/info/hannes/logcat/LogcatActivity.kt [hunk 1]
```diff
@@ -30,10 +30,8 @@ class LogcatActivity : AppCompatActivity() {
             // firstFragment.arguments = intent.extras
 
             // Add the fragment to the 'fragment_container' FrameLayout
-            supportFragmentManager.beginTransaction()
-                    .add(R.id.fragment_container, firstFragment).commit()
```
Comment (hannesa2) [inline LogcatCoreLib/src/main/java/info/hannes/logcat/LogcatActivity.kt:34]: In general, to avoid too much noise, I recommend to separate such changes in an own commit, the reduce following commits to real changes

File: LogcatCoreLib/src/main/java/info/hannes/logcat/LogcatActivity.kt [hunk 2]
```diff
@@ -30,10 +30,8 @@ class LogcatActivity : AppCompatActivity() {
...
+            supportFragmentManager.beginTransaction().add(R.id.fragment_container, firstFragment).commit()
         }
-
```
Comment (hannesa2) [inline LogcatCoreLib/src/main/java/info/hannes/logcat/LogcatActivity.kt:36]: same here

File: LogcatCoreLib/src/main/java/info/hannes/logcat/LogcatFragment.kt [hunk 3]
```diff
@@ -49,6 +48,7 @@ class LogcatFragment : LogBaseFragment() {
     }
 
     companion object {
+
```
Comment (hannesa2) [inline LogcatCoreLib/src/main/java/info/hannes/logcat/LogcatFragment.kt:51]: sometimes add a line

File: LogcatCoreLib/src/main/java/info/hannes/logcat/LogcatFragment.kt [hunk 4]
```diff
@@ -58,6 +58,5 @@ class LogcatFragment : LogBaseFragment() {
             fragment.arguments = args
             return fragment
         }
-
```
Comment (hannesa2) [inline LogcatCoreLib/src/main/java/info/hannes/logcat/LogcatFragment.kt:61]: Sometimes remove a line makes only noise and should be separated in a format commit.
I'm not speaking about necessary of it.

File: LogcatCoreLib/src/main/java/info/hannes/logcat/LogfileFragment.kt [hunk 5]
```diff
@@ -10,30 +10,21 @@ import info.hannes.timber.FileLoggingTree
 import info.hannes.timber.fileLoggingTree
 import java.io.File
 
-class LogfileFragment : LogBaseFragment(), Observer<String> {
+class LogfileFragment : LogBaseFragment(), Observer<Event<String>> {
```
Comment (hannesa2) [inline LogcatCoreLib/src/main/java/info/hannes/logcat/LogfileFragment.kt:13]: The `Event`  looks like a debounce. The purpose It's not understandable on the name itself. But I've no better suggestion

File: LogcatCoreLib/src/main/java/info/hannes/logcat/LogfileFragment.kt [hunk 6]
```diff
@@ -10,30 +10,21 @@ import info.hannes.timber.FileLoggingTree
...
-    override fun onStart() {
-        super.onStart()
-        FileLoggingTree.lastLogEntry.observeForever(this)
-    }
-
-    override fun onStop() {
-        FileLoggingTree.lastLogEntry.removeObserver(this)
-        super.onStop()
+        val view = super.onCreateView(inflater, container, savedInstanceState)
+        FileLoggingTree.lastLogEntry.observe(viewLifecycleOwner, this)
```
Comment (hannesa2) [inline LogcatCoreLib/src/main/java/info/hannes/logcat/LogfileFragment.kt:19]: Hmm, now you observe it always. I guess because of the duplicate-last-line-issue. I'm not 100 % convinced

File: LogcatCoreLib/src/main/java/info/hannes/logcat/base/LogBaseFragment.kt [hunk 7]
```diff
@@ -26,7 +27,7 @@ import java.io.File
 abstract class LogBaseFragment : Fragment() {
 
     private var verboseItem: MenuItem? = null
-    private lateinit var logsRecycler: RecyclerView
+    private var logsRecycler: RecyclerView? = null
```
Comment (hannesa2) [inline LogcatCoreLib/src/main/java/info/hannes/logcat/base/LogBaseFragment.kt:30]: Please avoid this null ability here and switch back to `lateinit`

**Review by hannesa2 — CHANGES_REQUESTED** [review]

File: LogcatCoreLib/src/main/java/info/hannes/logcat/LogfileFragment.kt (hunk 6 above)
Comment (1951FDG) [reply LogcatCoreLib/src/main/java/info/hannes/logcat/LogfileFragment.kt:19]: But observation is tied to lifecycle, so it's not always observed, maybe this article sheds more light, https://medium.com/@cs.ibrahimyilmaz/viewlifecycleowner-vs-this-a8259800367b

File: LogcatCoreLib/src/main/java/info/hannes/logcat/LogfileFragment.kt (hunk 5 above)
Comment (1951FDG) [reply LogcatCoreLib/src/main/java/info/hannes/logcat/LogfileFragment.kt:13]: I used the same class name as Google, I decided to leave the class intact so others can relate to it as well in terms of web familiarity, don't know if I'm making sense here, https://github.com/android/architecture-samples/blob/main/app/src/main/java/com/example/android/architecture/blueprints/todoapp/Event.kt

File: LogcatCoreLib/src/main/java/info/hannes/logcat/LogcatActivity.kt (hunk 1 above)
Comment (1951FDG) [reply LogcatCoreLib/src/main/java/info/hannes/logcat/LogcatActivity.kt:34]: I don't understand fully, I grouped all the format changes in one commit, should I have made several commits, one for each file?

**Comment by 1951FDG** [comment]
> Thank you !
> 
> I know this wish (not a request !) is maybe too much, but it would be super nice, to have a test in advance which fails before the fix and the fix solves it.
> eg. the duplicate last line I never observed

I don't know how to write a test for this to be honest, I can however make a video, before and after for every commit that fixes a bug, that you can watch in slow motion, may that be of help? I'll be happy to do this, or maybe there is another way?

I separated the commits in such a way, very carefully, that you could checkout before and after, for all these bugs, the fragment you need to test is `Timber logfile`, **IMPORTANT** **_I think I forgot to mention this._**

For example you could checkout commit before 8cfe21e to observe bug by going to activity, then press square icon in navigation bar (used to switch apps), then go back to activity instead of switching apps.

File: LogcatCoreLib/src/main/java/info/hannes/logcat/base/LogBaseFragment.kt (hunk 7 above)
Comment (1951FDG) [reply LogcatCoreLib/src/main/java/info/hannes/logcat/base/LogBaseFragment.kt]: I've add a commit that reverts back to lateinit

Comment (hannesa2) [reply LogcatCoreLib/src/main/java/info/hannes/logcat/base/LogBaseFragment.kt]: Great, but ...

I know this. There comes a requested change, I do it quickly and introduce an error. 

image

This doesn't work and introduce an error. The solution is to revert this line complete

Comment (1951FDG) [reply LogcatCoreLib/src/main/java/info/hannes/logcat/base/LogBaseFragment.kt]: Sorry, I pushed this last change too quickly, I've pushed another commit to fix this. I'm in the middle of moving...

**Comment by hannesa2** [comment]
Please have a look on conflicting files

**Comment by 1951FDG** [comment]
I'll be back on this sometime next week. thanks.

**Comment by hannesa2** [comment]
Concerning duplicate entries I made an Espresso test #173 to check this.
Off course it detects the duplicate entries but I disabled it
https://github.com/AppDevNext/Logcat/pull/173/files#diff-7b66b5b6bb25f12bb9e9074ed8c48ba1ea6189965a225a77f6acdd111b17ecffR31

I extracted this fix in an own pull request #168 and there I removed this `@Ignore`

**Comment by hannesa2** [comment]
Please can you rebase it ? Then I see easy, what's obsolete now
